File: java/evaluate_reverse_polish_notation.java
public class Solution {
    boolean isOp(String s) {
        return s.equals("+") || s.equals("-") || s.equals("*") || s.equals("/");
    }
    
    int calc(int a, int b, String op) {
        if (op.equals("+")) return a+b;
        else if (op.equals("-")) return a-b;
        else if (op.equals("*")) return a*b;
        else if (op.equals("/")) return a/b;
        else return -1;
    }
    
    public int evalRPN(String[] tokens) {
        List<Integer> nums=new ArrayList<>();
        for (int i=0; i<tokens.length; i++) {
            if (isOp(tokens[i])) {
                if (nums.size()<2) return -1;
                int b=nums.get(nums.size()-1);
                nums.remove(nums.size()-1);
                int a=nums.get(nums.size()-1);
                nums.remove(nums.size()-1);
                nums.add(calc(a, b, tokens[i]));
            } else {
                nums.add(Integer.parseInt(tokens[i]));
            }
        }    
        return nums.get(nums.size()-1);
    }
}

File: cpp/evaluate_reverse_polish_notation.cpp
150. Evaluate Reverse Polish Notation  QuestionEditorial Solution  My Submissions
Total Accepted: 71093 Total Submissions: 288951 Difficulty: Medium
Evaluate the value of an arithmetic expression in Reverse Polish Notation.

Valid operators are +, -, *, /. Each operand may be an integer or another expression.

Some examples:
  ["2", "1", "+", "3", "*"] -> ((2 + 1) * 3) -> 9
  ["4", "13", "5", "/", "+"] -> (4 + (13 / 5)) -> 6
Hide Company Tags LinkedIn
Hide Tags Stack
Hide Similar Problems (H) Basic Calculator (H) Expression Add Operators

class Solution {
public:
    bool isop(string s) {
        if (s=="+" || s=="-" || s=="*" || s=="/") return true;
        return false;
    }
    
    int calc(int num1, int num2, string op) {
        if (op=="+") return num1+num2;
        else if (op=="-") return num1-num2;
        else if (op=="*") return num1*num2;
        else if (op=="/") return num1/num2;
        else return 0;
    }
    
    int evalRPN(vector<string>& tokens) {
        stack<int> stk;
        for (auto n:tokens) {
            if (!isop(n)) stk.push(stoi(n));
            else {
                int num2=stk.top();
                stk.pop();
                int num1=stk.top();
                stk.pop();
                stk.push(calc(num1, num2, n));
            }
        }
        return stk.top();
    }
};

